src/net/server/handlers/mul.rs

```diff
@@ -192,16 +197,243 @@ mod tests {
 
         // try a request 10 times
         for _ in 0..10 {
-            send_req(
-                &mut rx,
-                port,
-                query_id,
-                step.clone(),
-                target_helper,
-                offset,
-                body,
-            )
-            .await;
+            let resp = send_req(port, query_id, step.clone(), target_helper, offset, body).await;
+
+            let status = resp.status();
+            let resp_body = body::to_bytes(resp.into_body()).await.unwrap();
+            let resp_body_str = String::from_utf8_lossy(&resp_body);
+
+            // response comparison
+            let channel_id = ChannelId {
+                identity: target_helper,
+                step: step.clone(),
+            };
+            let env = [0; DATA_SIZE as usize].to_vec().into_boxed_slice();
+            let envs = (0..DATA_LEN)
+                .map(|i| MessageEnvelope {
+                    record_id: i.into(),
+                    payload: env.clone(),
+                })
+                .collect::<Vec<_>>();
+
+            assert_eq!(status, StatusCode::OK, "{}", resp_body_str);
+            let messages = rx.try_recv().expect("should have already received value");
+            assert_eq!(messages, (channel_id, envs));
+        }
+    }
+
+    #[allow(clippy::mutable_key_type)] // `HeaderName` is known good key
+    fn build_malformed_req<Q: Into<String>, S: Into<String>, I: Into<String>, B: Into<Body>>(
+        port: u16,
+        query_id: Q,
+        step: S,
+        identity: I,
+        headers: HashMap<HeaderName, HeaderValue>,
+        body: B,
+    ) -> Request<Body> {
+        let uri = format!(
+            "http://127.0.0.1:{}/mul/query-id/{}/step/{}?identity={}",
+            port,
+            query_id.into(),
+            step.into(),
+            identity.into(),
+        );
+
+        let mut req = Request::post(uri);
+        let req_headers = req.headers_mut().unwrap();
+        for (key, value) in headers {
+            req_headers.insert(key, value);
+        }
+        req.body(body.into()).unwrap()
+    }
+
+    async fn expect_res(req: Request<Body>, expected_status: StatusCode) {
+        let resp = Client::default()
+            .request(req)
+            .await
+            .expect("request should complete successfully");
+        assert_eq!(resp.status(), expected_status);
+    }
+
+    #[tokio::test]
+    #[allow(clippy::mutable_key_type)] // [`HeaderName`] is known good key
+    #[allow(clippy::too_many_lines)] // testing all permutations of request
```

Comment:
I agree with clippy here - you could split this test into multiple by creating a helper function that prepares a valid request. then each test will replace one part of that request with the malformed entity and execute it

Reply:
In addition to be more readable, this will also be faster because rust executes tests in parallel

Reply:
ok, i came up with a much cleaner version of this test.